src/overlays/actors/ovl_En_Am/z_en_am.c

```diff
@@ -93,63 +112,881 @@ static ColliderQuadInit D_809AFFD8 = {
+    /* Giant's Knife */ DMG_ENTRY(4, AM_DMGEFF_KILL),
+    /* Fire arrow    */ DMG_ENTRY(2, AM_DMGEFF_KILL),
+    /* Ice arrow     */ DMG_ENTRY(4, AM_DMGEFF_ICE),
+    /* Light arrow   */ DMG_ENTRY(2, AM_DMGEFF_KILL),
+    /* Unk arrow 1   */ DMG_ENTRY(2, AM_DMGEFF_NONE),
+    /* Unk arrow 2   */ DMG_ENTRY(2, AM_DMGEFF_NONE),
+    /* Unk arrow 3   */ DMG_ENTRY(2, AM_DMGEFF_NONE),
+    /* Fire magic    */ DMG_ENTRY(0, AM_DMGEFF_MAGIC_FIRE_LIGHT),
+    /* Ice magic     */ DMG_ENTRY(3, AM_DMGEFF_ICE),
+    /* Light magic   */ DMG_ENTRY(0, AM_DMGEFF_MAGIC_FIRE_LIGHT),
+    /* Shield        */ DMG_ENTRY(0, AM_DMGEFF_NONE),
+    /* Mirror Ray    */ DMG_ENTRY(0, AM_DMGEFF_NONE),
+    /* Kokiri spin   */ DMG_ENTRY(1, AM_DMGEFF_NONE),
+    /* Giant spin    */ DMG_ENTRY(4, AM_DMGEFF_KILL),
+    /* Master spin   */ DMG_ENTRY(2, AM_DMGEFF_KILL),
+    /* Kokiri jump   */ DMG_ENTRY(2, AM_DMGEFF_NONE),
+    /* Giant jump    */ DMG_ENTRY(8, AM_DMGEFF_KILL),
+    /* Master jump   */ DMG_ENTRY(4, AM_DMGEFF_KILL),
+    /* Unknown 1     */ DMG_ENTRY(0, AM_DMGEFF_NONE),
+    /* Unblockable   */ DMG_ENTRY(0, AM_DMGEFF_NONE),
+    /* Hammer jump   */ DMG_ENTRY(4, AM_DMGEFF_KILL),
+    /* Unknown 2     */ DMG_ENTRY(0, AM_DMGEFF_NONE),
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Am/func_809AEFA4.s")
+static InitChainEntry sInitChain[] = {
+    ICHAIN_S8(naviEnemyId, 19, ICHAIN_CONTINUE),
+    ICHAIN_F32_DIV1000(gravity, -4000, ICHAIN_CONTINUE),
+    ICHAIN_F32(targetArrowOffset, 5300, ICHAIN_STOP),
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_En_Am/func_809AF0DC.s")
+void EnAm_SetupAction(EnAm* this, EnAmActionFunc actionFunc) {
+    this->actionFunc = actionFunc;
+}
+
+/**
+ * Checks bgCheckFlags in the direction of current yaw at the specified distance.
+ *
+ * Returns true if the armos would land on the ground in the resulting position.
+ *
+ * If it won't land on the ground, it will return true anyway if the floor the armos will be on
+ * is 20 units higher than the home position. This is a convoluted way of making sure
+ * the Armos can move down slight slopes but not steep ones (or complete drop offs).
+ */
+s32 EnAm_CanMove(EnAm* this, GlobalContext* globalCtx, f32 distance, s16 yaw) {
+    s16 ret;
+    s16 curBgCheckFlags;
+    f32 sin;
+    f32 cos;
+    Vec3f curPos;
+
+    // save current position and bgCheckFlags to restore later
+    curPos = this->dyna.actor.world.pos;
+    curBgCheckFlags = this->dyna.actor.bgCheckFlags;
+
+    sin = Math_SinS(yaw) * distance;
+    cos = Math_CosS(yaw) * distance;
+
+    this->dyna.actor.world.pos.x += sin;
+    this->dyna.actor.world.pos.z += cos;
+
+    Actor_UpdateBgCheckInfo(globalCtx, &this->dyna.actor, 0.0f, 0.0f, 0.0f, 4);
+    this->dyna.actor.world.pos = curPos;
+    ret = this->dyna.actor.bgCheckFlags & 1;
+
+    if (!ret) {
+        if ((this->dyna.actor.home.pos.y - 20.0f) <= this->dyna.actor.floorHeight) {
+            ret = true;
+        }
+    }
```

Comment:
```suggestion
    if (!ret && (this->dyna.actor.floorHeight >= (this->dyna.actor.home.pos.y - 20.0f))) {
        ret = true;
    }
```
I think this is more clear.